### ISSUE TITLE
Versioning conda packages via git no longer works

### ISSUE THREAD

**Issue by jhunkeler** [conda/conda-build#713]
Hi,

It seems `1.18.2` has broken versioning for git builds. The wall of text below shows the build process of a simple package in both the latest release of `conda-build` as well as the previous release. 

The git repository I am building against has an annotated tag "v0.1" and is currently 25 commits ahead.

The expected output file: 
`specview-v0.1.git-py35_25.tar.bz2`

The actual output file: 
`specview-0.0.0.git-py35_0.tar.bz2`

I'm hoping I am just doing something wrong, or maybe the documentation hasn't been updated to reflect a new requirement for building with git. Either way, I've been stuck using `1.18.1` since `1.18.2` was released.

**Quick Notes:** 
* `conda-build`'s git master recipe produces the same result as the `1.18.2` release version. Just in case you wanted me to try it ;)
* In testing I modified the `environ.get()` call to match conda's documentation, but the result was the same. I've been building with these `GIT_` references for several months without encountering any issues.
* This appears to be OS agnostic. The results are the same on OS X and Linux.

## meta.yaml

``` yaml
about:
    home: https://github.com/spacetelescope/specview
    license: BSD
    readme: README.rst
    summary: specview
build:
    number: '{{ environ.get("GIT_DESCRIBE_NUMBER", 0) }}'
package:
    name: specview
    version: '{{ environ.get("GIT_DESCRIBE_TAG", "0.0.0") }}.git'
requirements:
    build:
    - astropy
    - numpy
    - scipy
    - pyqtgraph
    - setuptools
    - python
    run:
    - astropy
    - numpy
    - scipy
    - pyqtgraph
    - python
source:
    git_url: https://github.com/spacetelescope/specview
```

## 1.18.2

```
BUILD START: specview-0.0.0.git-py35_0
Fetching package metadata:
Solving package specifications:
The following packages will be downloaded:

    package                    |            build
    ---------------------------|-----------------
    pyqtgraph-0.9.10           |           py35_0         890 KB

The following NEW packages will be INSTALLED:

    astropy:    1.1-np110py35_0
    freetype:   2.5.5-0
    libpng:     1.6.17-0
    numpy:      1.10.2-py35_0
    openssl:    1.0.2d-0
    pip:        7.1.2-py35_0
    pyopengl:   3.1.1a1-np110py35_0
    pyqt:       4.11.4-py35_1
    pyqtgraph:  0.9.10-py35_0
    python:     3.5.1-0
    qt:         4.8.7-1
    readline:   6.2-2
    scipy:      0.16.1-np110py35_0
    setuptools: 19.1.1-py35_0
    sip:        4.16.9-py35_0
    sqlite:     3.8.4.1-1
    tk:         8.5.18-0
    wheel:      0.26.0-py35_1
    xz:         5.0.5-0
    zlib:       1.2.8-0

Fetching packages ...
pyqtgraph-0.9. 100% |######################################| Time: 0:00:00  77.65 MB/s
Extracting packages ...
[      COMPLETE      ]|#########################################################| 100%
Linking packages ...
[      COMPLETE      ]|#########################################################| 100%
Removing old work directory
==> git log -n1 <==

commit 043d966c8ede31a2572d48f9fdc8e31067fefdfd
Author: busko <busko@stsci.edu>
Date:   Tue Jul 21 14:55:09 2015 -0400

    Fix bug related to bogus Qt signals.


==> git describe --tags --dirty <==

v0.1-25-g043d966


==> git status <==

On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean


Package: specview-0.0.0.git-py35_0
source tree in: /Users/jhunk/anaconda3/conda-bld/work
+ python setup.py install --offline
Freezing version number to specview/version.py
/Users/jhunk/anaconda3/envs/_build/lib/python3.5/site-packages/setuptools-19.1.1-py3.5.egg/setuptools/dist.py:284: UserWarning: Normalizing '0.2dev367' to '0.2.dev367'
[...]
Using /Users/jhunk/anaconda3/envs/_build/lib/python3.5/site-packages
Finished processing dependencies for specview==0.2.dev367
found egg dir: /Users/jhunk/anaconda3/envs/_build/lib/python3.5/site-packages/specview-0.2.dev367-py3.5.egg
number of files: 180
Fixing permissions
Detected hard-coded path in text file bin/sview
Fixing permissions
BUILD END: specview-0.0.0.git-py35_0
Nothing to test for: specview-0.0.0.git-py35_0
```

## 1.18.1

```
BUILD START: specview-v0.1.git-py35_25
Fetching package metadata: ......

WARNING: conda-build appears to be out of date. You have version 1.18.1 but the
latest version is 1.18.2. Run

conda update -n root conda-build

to get the latest version.

Solving package specifications: ............................
The following NEW packages will be INSTALLED:

    astropy:    1.1-np110py35_0
    freetype:   2.5.5-0
    libpng:     1.6.17-0
    numpy:      1.10.2-py35_0
    openssl:    1.0.2d-0
    pip:        7.1.2-py35_0
    pyopengl:   3.1.1a1-np110py35_0
    pyqt:       4.11.4-py35_1
    pyqtgraph:  0.9.10-py35_0
    python:     3.5.1-0
    qt:         4.8.7-1
    readline:   6.2-2
    scipy:      0.16.1-np110py35_0
    setuptools: 19.1.1-py35_0
    sip:        4.16.9-py35_0
    sqlite:     3.8.4.1-1
    tk:         8.5.18-0
    wheel:      0.26.0-py35_1
    xz:         5.0.5-0
    zlib:       1.2.8-0

Linking packages ...
[      COMPLETE      ]|###################################################################################################################################################| 100%
Removing old work directory
Cloning into '/Users/jhunk/anaconda3/conda-bld/work'...
done.
Submodule 'astropy_helpers' (https://github.com/astropy/astropy-helpers.git) registered for path 'astropy_helpers'
Cloning into 'astropy_helpers'...
remote: Counting objects: 2763, done.
remote: Compressing objects: 100% (13/13), done.
remote: Total 2763 (delta 6), reused 0 (delta 0), pack-reused 2750
Receiving objects: 100% (2763/2763), 753.12 KiB | 0 bytes/s, done.
Resolving deltas: 100% (1842/1842), done.
Checking connectivity... done.
Submodule path 'astropy_helpers': checked out '769d53eaf0d163d73d544a5ec060ee210420b1a0'
==> git log -n1 <==

commit 043d966c8ede31a2572d48f9fdc8e31067fefdfd
Author: busko <busko@stsci.edu>
Date:   Tue Jul 21 14:55:09 2015 -0400

    Fix bug related to bogus Qt signals.


==> git describe --tags --dirty <==

v0.1-25-g043d966


==> git status <==

On branch master
Your branch is up-to-date with 'origin/master'.
nothing to commit, working directory clean


Package: specview-v0.1.git-py35_25
source tree in: /Users/jhunk/anaconda3/conda-bld/work
+ python setup.py install --offline
Freezing version number to specview/version.py
/Users/jhunk/anaconda3/envs/_build/lib/python3.5/site-packages/setuptools-19.1.1-py3.5.egg/setuptools/dist.py:284: UserWarning: Normalizing '0.2dev367' to '0.2.dev367'
[...]
Using /Users/jhunk/anaconda3/envs/_build/lib/python3.5/site-packages
Finished processing dependencies for specview==0.2.dev367
found egg dir: /Users/jhunk/anaconda3/envs/_build/lib/python3.5/site-packages/specview-0.2.dev367-py3.5.egg
number of files: 180
Fixing permissions
Detected hard-coded path in text file bin/sview
Fixing permissions
BUILD END: specview-v0.1.git-py35_25
Nothing to test for: specview-v0.1.git-py35_25
```

**Comment by mjuric** [comment]
I'm seeing the same problem with [this meta.yaml](https://github.com/mjuric/conda-lsst/blob/master/recipes/static/eups/meta.yaml). The generated version is an empty string; e.g.:

```
mjuric@gallifrey:~/projects/lsst_packaging_conda$ conda build recipes/static/eups
Removing old build environment
Removing old work directory
BUILD START: eups--7
Fetching package metadata: ......
Solving package specifications: ..........
...
```

(note the `eups--7` part; EUPS is the product name, and 7 is the build number).

**Comment by scopatz** [comment]
I am also seeing this issue with xonsh while trying to get a release out. The meta.yaml file can be seen [here](https://github.com/scopatz/xonsh/blob/f7081ac85327e272536f15f61dc9ea0f212738cc/recipe/meta.yaml)

**Comment by msarahan** [comment]
pinging @stuarteberg - any idea here?

**Comment by stuarteberg** [comment]
Without examining the above logs too closely, my best guess is that this is an (almost) duplicate #708.  When I originally investigated that issue, I found that indeed, `1.18.2` is broken, but thankfully the latest version of `master` is not.

@jakirkham did some follow-up, and determined that #621 was to blame.  (That PR caused multiple unexpected bugs related to `git` support.)  Fortunately, it has already been reverted (via #694), so my recommendation is to simply release `1.18.3` as soon as possible.

If the reporters above (@jhunkeler, @mjuric, and @scopatz) would like to verify that their recipes work with the master branch of `conda-build`, that would be helpful.

**Comment by jhunkeler** [comment]
@stuarteberg @msarahan 

I confirm the master branch is working now:

```
WARNING: conda-build appears to be out of date. You have version 1.10alpha.0 but the
latest version is 1.18.2. Run

conda update -n root conda-build

to get the latest version.

[...]

Nothing to test for: specview-v0.1.git-py35_25
# If you want to upload this package to anaconda.org later, type:
#
# $ anaconda upload /home/jhunk/anaconda3/conda-bld/linux-64/specview-v0.1.git-py35_25.tar.bz2

```

**Comment by msarahan** [comment]
Pinging @ilanschnell and @groutr - when can we schedule the next conda-build release?

**Comment by ilanschnell** [comment]
I think we should schedule the next conda-build release for January 11, 2016.

**Comment by jakirkham** [comment]
Sorry just got back from a flight and am getting over a cold. I think @stuarteberg is a correct that this is a duplicate of this issue ( https://github.com/conda/conda-build/issues/708 ). This *should* be fixed in `master`.

**Comment by jhunkeler** [comment]
Just checking in. Is there anything currently blocking the 1.18.3 release?

**Comment by kalefranz** [comment]
@jhunkeler I'm the new conda tech lead, and everybody was probably waiting on me to start.  Lots of backlog to catch up on, and I hadn't even thought of needing a set of releases yet.  But it's possible.  I'll dig in on releases and see what can or is ready to go out.

**Comment by jhunkeler** [comment]
@kalefranz Not a problem! I appreciate your reply. The git master build still works for me, so I'm not dead in the water or anything! ;)

**Comment by meawoppl** [comment]
Is the current conda in the APD broken?  I have encountered this in our CI stack which pulls miniconda and installs conda-build.  

This is pressing for us.  Please let me know.

**Comment by jakirkham** [comment]
I don't follow, @meawoppl. Not only can I not tell that you are hitting this bug, but I cannot tell you are hitting any bug at all. It looks like your build completed successfully. If this is not the case, could you please point out the line where you are having a problem?

**Comment by meawoppl** [comment]
Ugh.  Ok.    Would you accept a PR of a test to make sure this dosen't crop up in future versions?

This appears to affect 1.19.0 (current release?)

**Comment by msarahan** [comment]
Absolutely!  Sorry for your troubles, though.

**Comment by jakirkham** [comment]
I believe this was already tested. I didn't realize a 1.19.0 release was just recently made. Did this actually get included? In which case, would you mind sharing your recipe @meawoppl?

**Comment by meawoppl** [comment]
This is blocking our production deploy atm, so I am trying to figure out what the right thing to do is.  Call out conda-build 1.18.1 specifically?  I don't know the date of the regression, but I can poke through our CI server and try to narrow it down.  I will come in with a PR later after I finish our scheduled downtime (Wednesday?).

**Comment by meawoppl** [comment]
Basically, it won't update to 0.0.0, which I consider sane behavior.  That said, its the only version our CI will write atm.

**Comment by jakirkham** [comment]
So, I have rechecked with the OPs recipe and actually conda-build 1.19.0 works fine. I was able to build specview and it correctly updated the version to `v0.1` after the initial recipe parsing passes. So the final package was named as so `specview-v0.1.git-py27_25.tar.bz2`.

Again @meawoppl please share your recipe. I would like to help you, but without more information about what you are doing I am unable to.

**Comment by meawoppl** [comment]
Here is a minimum reproduction case:
https://github.com/meawoppl/conda-git-regression

Here is the conda build output:
http://pastebin.com/DePfCPBw

**Comment by jakirkham** [comment]
Thank you. I'll take a look.

**Comment by jakirkham** [comment]
So, the problem is you are using `setup.py` for the build step. In order to make this work, you need to write a `build.sh` and a `build.bat` file.

A simple example with `conda` itself can be seen here. ( https://github.com/conda/conda-recipes/tree/master/python/conda ) This should be pretty much copy and paste if you are dealing with a standard python package. This often works for more complex python packages as well.

If you have more questions, I am more than willing to help. Hope that helps.

**Comment by jakirkham** [comment]
Though that is pretty special how `script` is being handled here. That is probably worth a new issue. If you wish to open one @meawoppl, I think that would be a good idea. Just include everything from your last comment. This is a different, albeit related, problem. Adding `build.sh` and `bld.bat` will allow you to workaround this.

**Comment by meawoppl** [comment]
Interesting.  This was written to be representative of our rough template for packages.  In our meta.yaml we callout the "script", but in the packages we also have build.bat and build.sh.  I think this is left over from here: 
https://github.com/conda/conda-build/issues/270

Which of these takes priority, and has this changed as a result of the other work done in this space?

**Comment by meawoppl** [comment]
Interesting.  Adding the files does seem to change things for the better.  Should I take out my call-outs to 'script' if I have these files?

**Comment by msarahan** [comment]
Maybe as a workaround, but this sounds like a bug.  If it's a documented feature that isn't working correctly, then it needs fixing.  If you have a reproducible test case, I think we should take a look.

**Comment by jakirkham** [comment]
So, the OPs bug here is a *fun* story. Somebody made a well meaning change to refactor some code. However, there are some corner cases they missed so it didn't behave correctly. That caused the bug seen here. When trying to fix some other bugs, we hit this one in some of our tests. It was difficult to figure out what exactly went wrong so we simply ripped out the old patch. Why this still affects `script` seems to be a different story that should be investigated.

While #270 may also be a problem, I ran into the same issue as you on a Linux machine. So, I don't think that is only a Windows problem. Still worthy of a new issue IMHO.

Personally, I stick with `build.sh` and `bld.bat` as these tend to make up the norm for recipe building cases. If something goes wrong with them, it tends to get fixed reasonably fast. For the meantime, this may be reasonable for you too. However, you are right that it is uncool that `script` isn't behaving correctly.

**Comment by jakirkham** [comment]
>  In our meta.yaml we callout the "script", but in the packages we also have build.bat and build.sh.
> ...
> Which of these takes priority, and has this changed as a result of the other work done in this space?

Sorry, failed to answer your other point, @meawoppl. I would say pick one of the two strategies. It's not clear to me whether there is defined behavior when both are present.

**Comment by meawoppl** [comment]
I was mistakenly using the 1.18.0 version of conda

```
Copying /home/meawoppl/repos/3scan/conda-git-regression to /home/meawoppl/anaconda3/conda-bld/work
Package: 3scan-shared-0.0.0-py34_0
source tree in: /home/meawoppl/anaconda3/conda-bld/work
+ python setup.py install
running install
running bdist_egg
running egg_info
creating DemoDemo.egg-info
writing DemoDemo.egg-info/PKG-INFO
writing dependency_links to DemoDemo.egg-info/dependency_links.txt
writing top-level names to DemoDemo.egg-info/top_level.txt
writing manifest file 'DemoDemo.egg-info/SOURCES.txt'
reading manifest file 'DemoDemo.egg-info/SOURCES.txt'
writing manifest file 'DemoDemo.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
running build_py
creating build
creating build/lib
creating build/lib/Demo
copying Demo/__init__.py -> build/lib/Demo
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/Demo
copying build/lib/Demo/__init__.py -> build/bdist.linux-x86_64/egg/Demo
byte-compiling build/bdist.linux-x86_64/egg/Demo/__init__.py to __init__.cpython-34.pyc
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying DemoDemo.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying DemoDemo.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying DemoDemo.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying DemoDemo.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
zip_safe flag not set; analyzing archive contents...
creating dist
creating 'dist/DemoDemo-1.2.3-py3.4.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing DemoDemo-1.2.3-py3.4.egg
Copying DemoDemo-1.2.3-py3.4.egg to /home/meawoppl/anaconda3/envs/_build/lib/python3.4/site-packages
Adding DemoDemo 1.2.3 to easy-install.pth file

Installed /home/meawoppl/anaconda3/envs/_build/lib/python3.4/site-packages/DemoDemo-1.2.3-py3.4.egg
Processing dependencies for DemoDemo==1.2.3
Finished processing dependencies for DemoDemo==1.2.3
found egg: /home/meawoppl/anaconda3/envs/_build/lib/python3.4/site-packages/DemoDemo-1.2.3-py3.4.egg
number of files: 2
Fixing permissions
Fixing permissions
BUILD END: 3scan-shared-0.0.0-py34_0
Nothing to test for: 3scan-shared-0.0.0-py34_0
# If you want to upload this package to anaconda.org later, type:
#
# $ anaconda upload /home/meawoppl/anaconda3/conda-bld/linux-64/3scan-shared-0.0.0-py34_0.tar.bz2
#
# To have conda build upload to anaconda.org automatically, use
# $ conda config --set anaconda_upload yes

meawoppl@lena:~/repos/3scan/conda-git-regression$ git describe --tags --long
1.2.3-3-g7faadfe
meawoppl@lena:~/repos/3scan/conda-git-regression$ conda build --version
conda-build 1.19.0
```

This still appears to be an issue.

**Comment by meawoppl** [comment]
Demo repository now has bld.bat and build.sh, and still seems to demonstrate regression.

**Comment by jakirkham** [comment]
It also still has the `script` line. Does the problem persist when you remove this?

**Comment by jakirkham** [comment]
Ok, so I must have missed on my first read through but to set up git you need to use `git_url` not `path. Changing this fixes it without looking at any other change (even if you still use `script`). In short, the problem was just this line in the recipe. Sorry I missed it earlier.

**Comment by meawoppl** [comment]
Alright, I am going to give this a try.  This will not work for private git repos then I suppose?

**Comment by jakirkham** [comment]
Sounds good. Feel free to come back if you run into more issues.

> This will not work for private git repos then I suppose?

Shouldn't be a problem. Under the surface, `conda` is just calling `git` so anything you can do with `git` can be done here. You just might need to set up some ssh keys with your copy of `git` and use the `git://` syntax when your recipe is not packaged with your repo. If the recipe is in the repo (like your example), it will be no different. Also, you can try to use `git_tag` (or `git_rev`; both do the same thing) to specify a particular version or commit. By default, it refers to `HEAD`.

**Comment by meawoppl** [comment]
Alright, so if I have a meta.yaml in a private repo I have to specify this git_url as .  

Is this something that needs to be added to the docs somewhere?

**Comment by jasongrout** [comment]
+1 to adding it to the docs, and documenting it as a breaking change. I just tracked our broken packages down to this too.  It looks like it was this change: https://github.com/conda/conda-build/commit/7689ae1 (and then modified in this change: https://github.com/conda/conda-build/commit/fddfb32).  Is conda using semantic versioning? If so, I would say this is a backwards-incompatible change and the version number should be incremented appropriately.

**Comment by jasongrout** [comment]
By the way, we saw in debugging this problem that those two changes actually also break `python -m conda_build.environ`, since they don't check to see if `m` is `None`.

**Comment by jakirkham** [comment]
> Alright, so if I have a meta.yaml in a private repo I have to specify this git_url as .
> Is this something that needs to be added to the docs somewhere?

It is. Sorry, should have linked you to them. Felt like it was just simpler to write.

> The `git_url` can also be a relative path to the recipe directory

( http://conda.pydata.org/docs/building/meta-yaml.html#source-from-git )

**Comment by meawoppl** [comment]
Interesting.  I have a final (possibly related) note, which is that 
```conda build --output .```
Still produces output with the wrong version number.  I can not reporduce this locally, but it functions this way in out CI server.  More soon.

```
/home/ubuntu/miniconda3/conda-bld/noarch/3scan-shared-0.0.0-py_0.tar.bz2
(venv-system)ubuntu@box315:~/3scan-shared$ conda build --output .
/home/ubuntu/miniconda3/conda-bld/noarch/3scan-shared-0.0.0-py_0.tar.bz2
(venv-system)ubuntu@box315:~/3scan-shared$ conda build --version
conda-build 1.19.0
(venv-system)ubuntu@box315:~/3scan-shared$ git describe --tags --long
0.5.0-114-gd992a8d
```

**Comment by jakirkham** [comment]
Looks like a bug. Probably should be split out. Though I don't know if it ever worked for these sorts of recipes honestly.

**Comment by meawoppl** [comment]
I can transition to scraping the output of conda build ., but that seems frankly hackey.  Should I open  a separate issue on this?

**Comment by jakirkham** [comment]
> I can transition to scraping the output of conda build ., but that seems frankly hackey.

Sorry, I have no silver bullets for you. Could tell me why you need this information though? Maybe there is another way to get the same effect.

> Should I open a separate issue on this?

Yes, please do.

**Comment by meawoppl** [comment]
After building the package we use conda build --output . to figure out where the package file ends up (and what it is named) from there we move it to our conda repository based on the git branch.  This allows us to do builds on different branches and use conda to install to testing/production and other deployed sites we maintain.  

Is there a way to specify the directory to save it to?  If so I can make a tempfolder and copy-pasta it in the right direction.

**Comment by pitrou** [comment]
Same problem here. This seems to have been broken in the 1.18.2 to 1.19.0 transition.

**Comment by pitrou** [comment]
> Could you tell me why you need this information though?

We use an authentication token to `anaconda upload` our built packages. `conda build` doesn't seem to have a command-line option to pass the token.

**Comment by jakirkham** [comment]
Ok, I was about to asking which problem you, @pitrou, are having as there are a few different ones that have surfaced that all appear orthogonal to the original issue and fix. Thanks for clarifying. Also, thanks for clarifying that this is, in fact, a regression. Would you mind opening a new issue for the `conda build --output` and ping myself and @meawoppl when you do? Thanks.

**Comment by jakirkham** [comment]
Also, @jasongrout, do you think it is worth opening an issue regarding the `path` break and another for the `python -m conda_build.environ`? The latter definitely seems to need an issue. The former is probably a more general question about breaking changes and how to make and address the effect thereof (unless I am misunderstanding the problem here).

**Comment by jhunkeler** [comment]
@jakirkham speaking of conda_build.environ I filed #751 last night. I wasn't sure if it was directly related to this PR or not so I broke it out.

**Comment by jakirkham** [comment]
Perfect, thanks @jhunkeler. Also, do you find that your original problem here was solved? If so, would you mind closing this.

As for any number of other errors that might be related to `GIT_*` variables, I would like to politely recommend people open them as a new issue. This way it will be easier to follow what is going on.

**Comment by jhunkeler** [comment]
```
$ conda build --version
conda-build 1.19.0
```

And yes, it seems the original issue was addressed in `1.19.0`

```
$ conda build specview
[...]
BUILD END: specview-v0.1.git-py35_25
Nothing to test for: specview-v0.1.git-py35_25
```

**Comment by jakirkham** [comment]
Perfect, thanks again.

**Comment by jasongrout** [comment]
I think it's worth it to at least open an issue about documenting the breaking change regarding the path option, maybe in the changelog for now. It also makes sense to have the larger conversation about how to deal with breaking changes too, of course.
